File: benchmarks/tsb/bench_groupby_first.ts
import { DataFrame, DataFrameGroupBy } from "tsb";
const N = 100_000;
const keys = ["A", "B", "C", "D", "E"];
const df = new DataFrame({
  key: Array.from({ length: N }, (_, i) => keys[i % keys.length]),
  val: Array.from({ length: N }, (_, i) => i * 0.5),
  val2: Array.from({ length: N }, (_, i) => i % 100),
});
const gbObj = new DataFrameGroupBy(df, ["key"]);
const WARMUP = 3;
const ITERS = 20;
for (let i = 0; i < WARMUP; i++) gbObj.first();
const t0 = performance.now();
for (let i = 0; i < ITERS; i++) gbObj.first();
const total = performance.now() - t0;
console.log(JSON.stringify({ function: "groupby_first", mean_ms: total / ITERS, iterations: ITERS, total_ms: total }));

File: benchmarks/pandas/bench_groupby_first.py
import pandas as pd, time, json
N = 100_000
keys = ["A", "B", "C", "D", "E"]
df = pd.DataFrame({
    "key": [keys[i % len(keys)] for i in range(N)],
    "val": [i * 0.5 for i in range(N)],
    "val2": [i % 100 for i in range(N)],
})
gb = df.groupby("key")
WARMUP = 3
ITERS = 20
for _ in range(WARMUP):
    gb.first()
t0 = time.perf_counter()
for _ in range(ITERS):
    gb.first()
total = (time.perf_counter() - t0) * 1000
print(json.dumps({"function": "groupby_first", "mean_ms": total / ITERS, "iterations": ITERS, "total_ms": total}))
